controllers/hvpa_controller.go

```diff
@@ -1281,6 +1283,150 @@ func getThreshold(thresholdVals *autoscalingv1alpha1.ChangeParams, resourceType
+			applied := lastApplied.VpaStatus.ContainerResources[j]
+
+			if container.Name == applied.ContainerName && !reflect.DeepEqual(container.Resources, applied.Resources) {
+				vpaScaled = true
+				container.Resources = applied.Resources
+			}
+		}
+	}
+
+	switch kind {
+	case "Deployment":
+		deploy.Spec.Replicas = &replicas
+		if vpaScaled {
+			newPodSpec.DeepCopyInto(&deploy.Spec.Template.Spec)
+		}
+		newObj = deploy
+	case "StatefulSet":
+		ss.Spec.Replicas = &replicas
+		if vpaScaled {
+			newPodSpec.DeepCopyInto(&ss.Spec.Template.Spec)
+		}
+		newObj = ss
+	case "DaemonSet":
+		if vpaScaled {
+			newPodSpec.DeepCopyInto(&ds.Spec.Template.Spec)
+		}
+		newObj = ds
+	case "ReplicaSet":
+		rs.Spec.Replicas = &replicas
+		if vpaScaled {
+			newPodSpec.DeepCopyInto(&rs.Spec.Template.Spec)
+		}
+		newObj = rs
+	case "ReplicationController":
+		rc.Spec.Replicas = &replicas
+		if vpaScaled {
+			newPodSpec.DeepCopyInto(&rc.Spec.Template.Spec)
+		}
+		newObj = rc
+	default:
+		err := fmt.Errorf("TargetRef kind not supported %v", kind)
+		log.Error(err, "Error")
+		return err
+	}
+
+	if hpaScaled || vpaScaled {
+		log.V(2).Info("Scaling required. Applying last applied", "hvpa", hvpa)
+		return r.Update(context.TODO(), newObj)
+	}
+	log.V(2).Info("Scaling not required", "hvpa", hvpa)
+	return nil
+}
+
+func getScalingStatusFrom(hpaStatus *autoscaling.HorizontalPodAutoscalerStatus, vpaStatus *vpa_api.VerticalPodAutoscalerStatus, podSpec *corev1.PodSpec) *autoscalingv1alpha1.ScalingStatus {
+	scalingStatus := &autoscalingv1alpha1.ScalingStatus{}
+
+	if hpaStatus != nil {
+		scalingStatus.HpaStatus = autoscalingv1alpha1.HpaStatus{
+			CurrentReplicas: hpaStatus.CurrentReplicas,
+			DesiredReplicas: hpaStatus.DesiredReplicas,
+		}
+	}
+
+	if vpaStatus != nil && vpaStatus.Recommendation != nil {
+		podResources := autoscalingv1alpha1.VpaStatus{
+			ContainerResources: make([]autoscalingv1alpha1.ContainerResources, len(vpaStatus.Recommendation.ContainerRecommendations)),
+		}
+		for idx := range vpaStatus.Recommendation.ContainerRecommendations {
+			recommendation := &vpaStatus.Recommendation.ContainerRecommendations[idx]
+			podResources.ContainerResources[idx].ContainerName = recommendation.ContainerName
+			podResources.ContainerResources[idx].Resources = corev1.ResourceRequirements{
+				Requests: recommendation.Target,
+			}
+		}
+		scalingStatus.VpaStatus = podResources
+
+		if podSpec != nil {
+			// If podSpec is passed, we want to capture "Resources" fully for containers for which VPA taget is available.
+			// The podResources.ContainerResources[i] is populated here only for the container for which VPA target is available.
+			for i := range scalingStatus.VpaStatus.ContainerResources {
+				for j := range podSpec.Containers {
+					container := podSpec.Containers[j]
+					if scalingStatus.VpaStatus.ContainerResources[i].ContainerName == container.Name {
+						scalingStatus.VpaStatus.ContainerResources[i].Resources = container.Resources
```

Comment:
Wouldn't this potentially override [this](https://github.com/gardener/hvpa-controller/pull/65/files#diff-6354ac486a904d10093c497f09f83a3eR1406)?

Reply:
Yes, this is the idea here.
Here we want to capture full details only for those containers for which VPA has provided recommendations (in its `status.target`). Please note that VPA recommendations can be turned off individually for containers.
Since vpa's `target` doesn't have `limits`, we depend on `podSpec`, and overwrite what was previously captured using `target`.
This enables hvpa-controller to control `resources` only for containers for which VPA is enabled, so that other containers can still be controlled from outside, ie, can be manually changed and will not be reconciled using hvpa's `lastApplied`

Reply:
But then it might container the resources for only the current VPA recommendation which may not recommend for every container. Isn't [`ProcessedRecommendations`](https://github.com/gardener/hvpa-controller/pull/65/files#diff-61c5bea64511bdaec1e64fb43fc1773cR282) supposed to contain the latest resources for all the containers so far recommended by VPA. Not just what came from the current recommendation?

Reply:
All the recommendations are kept in VPA's status. Even if there are new recommendations for specific containers in the status, the recommendations of other containers are not removed from the VPA's status. So we always have all - changed and unchanged - recommendations all the time in VPA's status.

> But then it might contain the resources for only the current VPA recommendation which may not recommend for every container.

Even if this is a possibility that for certain containers VPA recommendations were never populated in the VPA's status until a time, then there is anyway no way for hvpa-controller to capture them, or to control the `resources` of those containers. The recommendations for those containers will be captured as and when they are updated in VPA's status. This will be like turning the update mode of those specific container from "Off" to "Auto".
Also, as mentioned above, VPA keeps the older recommendations of other containers in its status.
I hope this clarifies.

Reply:
👍 I was thinking we could get the potentially missing values from the target pod template. But what you mentioned is fine.